Improve the default project theme

Calinou opened the pull request:
The new default project theme uses StyleBoxFlat extensively for a more modern design and better scalability to multiple resolutions.

SVG icons are now used in place of PNG icons. While this does not allow for true vector-based icon drawing (icons are still rasterized at load-time), this makes the design work easier for contributors and opens the door to vector drawing in the future (e.g. with polygons or SDFs).

Like for editor icons, the SVG header file is now built automatically when a SVG file is changed. This removing the need for running `make_header.py` manually.

The **Use Hidpi** project setting has been removed in favor of a **Default Theme Scale** project setting, which allows creating the default theme at a higher/lower scale than the default. This can be used when designing GUIs with a high base resolution to ensure crisp visuals.

The new theme has slightly different metrics, which may result in elements being shifted around when porting `3.x` projects (hence the `breaks compat` label).

**Note:** Unlike LineEdit, TextEdit will appear to have no visible focus outline because its focus outline is clipped by the main StyleBox's area (this is not about drawing order). We need to fix this in the TextEdit node, ideally before this PR is merged. See https://github.com/godotengine/godot/issues/32996.

This closes https://github.com/godotengine/godot/issues/25349 and closes https://github.com/godotengine/godot/issues/8636 (by implementing a more flexible mechanism).

## Preview

*On all screenshots, the first button is focused so that you can see how the focus outline looks.*

### Default resolution

| Before | After |
|--------|-------|
| ![Before](https://user-images.githubusercontent.com/180032/127786747-c56be4a6-84de-45cf-abd2-6fccb2aba6f9.png) | ![After](https://user-images.githubusercontent.com/180032/148452629-fc443a39-20cf-4ec3-861e-25bb023c2ef4.png) |

### Fullscreen (`canvas_items` stretch mode)

*The new theme scales much better to higher resolutions thanks to its StyleBoxFlat-based design.*

| Before | After |
|--------|-------|
| ![Before](https://user-images.githubusercontent.com/180032/127786748-fd7b12a1-d847-4096-bb86-e16bc2973b37.png) | ![After](https://user-images.githubusercontent.com/180032/148452624-b469616c-d894-47e2-b207-6093f1e76d3c.png) |

## TODO

- [ ] Generate fallback images when the NanoSVG module is disabled.
  - Right now, if you disable the NanoSVG module, "Invalid image" errors will be spammed when the default project theme is created.
- [ ] Print a warning message when compiling Godot with the NanoSVG module disabled.
  - Alternatively, make NanoSVG a hard requirement for compiling Godot.

pycbouh commented:
Something's wrong with the texture progress control or with its surrounding container?

Calinou commented:
> Something's wrong with the texture progress control or with its surrounding container?

It looks as intended – it's just that the higher spacing makes it not fit anymore within the container, so its edge looks flush with the panel instead of having a margin.

aaronfranke commented:
Overall this is really nice!

I'm not sure I like the scrollbars, and the other corners. Is there any program that has pointy edges on them?

![corners](https://user-images.githubusercontent.com/1646875/127797514-3996b510-446e-4032-9ec0-e710bee56e97.png)


It seems a bit bright. If we have white text we want to ensure the background is dark enough that people can easily read it. Can we make the background a nice even `#444`? (This was modified with GIMP)

![darker](https://user-images.githubusercontent.com/1646875/127797520-9d43ce51-2461-40b2-bf90-92787145413e.png)

Calinou commented:
> It seems a bit bright. If we have white text we want to ensure the background is dark enough that people can easily read it. Can we make the background a nice even `#444`? We could potentially go even darker. (This was modified with GIMP)

The panel's background is translucent, which is why it appears to be relatively bright against the clear color.

aaronfranke commented:
> The panel's background is translucent,

Shouldn't it be opaque by default?

Calinou commented:
> Shouldn't it be opaque by default?

I can make it opaque by default, but one reason I made it translucent by default is that it makes multiple overlapping panels distinguishable out of the box (since they don't have borders anymore).

ee0pdt commented:
The corners… is it possible to have rounded edges? I can’t think of a modern interface that has anything other these days. 

I don’t mean to prevent anyone using this style, but just want to know if it’s possible to create an alternative

Calinou commented:
> The corners… is it possible to have rounded edges? I can’t think of a modern interface that has anything other these days.

They were rounded in an earlier revision of this theme, but there are two issues with rounded corners in StyleBoxFlat:

- When you enable antialiasing, the StyleBoxFlat's edges become blurry due to how StyleBoxFlat antialiasing is implemented. This wasn't an issue at some point – it broke during the 3.1 -> 3.2 cycle and we never managed to find a good solution for it. There's an issue about this, but I can't find it right now.
- When you disable antialiasing, rounded corners look bad at higher resolutions. Beleved corners look much better since there are only 45-degree angles.

ee0pdt commented:
@Calinou that’s a fantastic reply. Thanks for the info

I guess that this is down to preference - and given the limitations I’d probably go with a one pixel radius and keep it simple. 

If we could solve the anti-aliasing problem…

But either way, could we get your answer in the docs with some info on how to change the corner if one wanted to try rounded corners that would be most helpful I feel.

Calinou commented:
> But either way, could we get your answer in the docs with some info on how to change the corner if one wanted to try rounded corners that would be most helpful I feel.

The default theme is [generated via C++ code](https://github.com/godotengine/godot/blob/737d3d31a6332c1f9d20b141b75c633953430c16/scene/resources/default_theme/default_theme.cpp), but I can probably add a few project settings to allow changing the corner radius, corner detail and antialiasing. I wouldn't go too far in the customization options though, as the default project theme is meant to be used for prototyping only.

aaronfranke commented:
> but I can probably add a few project settings to allow changing the corner radius, corner detail and antialiasing.

I don't think this is necessary. We should choose sane & clean defaults and go with the expectation that people will be using a custom theme on top of the default theme.

groud commented:
Great work, having something a little big more modern is great. 

Two points:
- Even if I like the fact we switch to a gray tone (it's indeed a lot better to fit various game themes), I kind of liked the fact that the theme was easy to recognize in the wild. Not a lot of projects are made using Godot so I think it's always useful to be able to recognize the default theme. Maybe we could find a balance by having a slight accent blue color somewhere? Likely on the buttons? I wonder if we could have a configurable accent color that could be applied to the whole theme easily (defaulting to a subtle blue).
- I think, like others, that I preferred the rounded corners. If it's not possible for now, I agree bevels are an acceptable workaround. The only place where it does not fit is the scrollbar, it looks a little bit aggressive and thus out of place. I would either make it look like other bevels or simply make it a simple rectangle (like most modern UIs now).

Anyway, 4.0 really needed such improvements. Great job!

Calinou commented:
> Maybe we could find a balance by having a slight accent blue color somewhere? Likely on the buttons? I wonder if we could have a configurable accent color that could be applied to the whole theme easily (defaulting to a subtle blue).

Unfortunately, I don't think it's worth the added maintenance effort. You have to remember that unlike the editor theme, the default project theme will be included in every project. Any change that we make to it has the potential to break compatibility with existing projects.

groud commented:
> Unfortunately, I don't think it's worth the added maintenance effort. You have to remember that unlike the editor theme, the default project theme is included in every project. Any change that we make to it has the potential to break compatibility with existing projects.

I don't get it. Why would we change that? My idea was more to make it a project settings having an impact on the default theme. Even if we change the default color value, the last value could be stored in the project settings and kept. The idea would be simply to allow a simple customization if the default blue color would not fit. 

That being said I don't know how feasible this is, maybe it would require another class inheriting the base theme? I don't know. 

> I toned down the corner radii for the scrollbars since I initially opened this PR. Did you look at the latest screenshots in OP?

Ah my bad, I did not see it. It's indeed better.

aaronfranke commented:
> @groud My idea was more to make it a project settings having an impact on the default theme. 

There isn't really a point to do this. Users can just define a custom theme instead of modifying the default / base theme.

Calinou commented:
> There isn't really a point to do this. Users can just define a custom theme instead of modifying the default / base theme.

I'm inclined to agree with this too. We can't expect this default project theme to be a "one size fits all" solution. If you want to develop a polished game, you *will* have to create your own theme eventually.

A good-looking default project theme is just here to give a better first impression while prototyping.

aaronfranke commented:
What if we inverted the brightness? Just an idea, not sure if it's better for general purpose usage. Here are some images lazily edited with GIMP. Vs old theme:

| Before | After |
|--------|-------|
| ![Before](https://user-images.githubusercontent.com/180032/127786748-fd7b12a1-d847-4096-bb86-e16bc2973b37.png) | ![After](https://user-images.githubusercontent.com/1646875/130275186-f8b18b7f-e1ed-4348-bb01-f8bd6b540943.png) |

Vs this PR's theme:

| Before | After |
|--------|-------|
| ![Before](https://user-images.githubusercontent.com/180032/127887869-64896fc1-1181-4a99-a38e-2af566ee94ee.png) | ![After](https://user-images.githubusercontent.com/1646875/130275186-f8b18b7f-e1ed-4348-bb01-f8bd6b540943.png) |

pycbouh commented:
I think it's hard to compare against a monotone gray background. It would be nice to see it over some game elements, so that alpha is more readable.

Calinou commented:
Rebased against the latest `master` branch. I fixed the `default_theme_scale` project setting to also scale the default font, but no other changes were made.

Calinou committed:
Improve the default project theme

The new default project theme uses StyleBoxFlat extensively for
a more modern design and better scalability to multiple resolutions.

SVG icons are now used in place of PNG icons. While this does not
allow for true vector-based icon drawing (icons are still rasterized
at load-time), this makes the design work easier for contributors
and opens the door to vector drawing in the future (e.g. with polygons
or SDFs).

Like for editor icons, the SVG header file is now built automatically
when a SVG file is changed. This removing the need for running
`make_header.py` manually (TODO).

The "Use Hidpi" project setting has been removed in favor of a
"Default Theme Scale" project setting, which allows creating the
default theme at a higher/lower scale than the default.
This can be used when designing GUIs with a high base resolution
to ensure crisp visuals.

Co-authored-by: Yuri Sizov <yuris@humnom.net>

pycbouh commented:
I've done additional work here to help Calinou, he should be able to merge my changes that add the remaining styles and icons.

While working on it, I've run into a few cases of weirdness in some controls, which I do not believe are related to the theming changes and should be fixed in another PR, but I'd still like to mention it here for now:

![godot windows tools 64_2021-12-31_20-17-29](https://user-images.githubusercontent.com/11782833/147834455-946d1641-20ab-4acc-8f94-29cba9d3de83.png)

In `FileDialog` a header subtitle is used, but it looks strange in default theme. Maybe we need to play around with it more, or maybe this dialog shouldn't use a header here at all. Also note the hidden files visibility icon. You can't tell if it's on or off. We should better distinguish the states, and I propose to use the same icons we have in the gizmo visibility/scene dock in the editor for the "off" state.

![godot windows tools 64_2021-12-31_20-20-09](https://user-images.githubusercontent.com/11782833/147834495-78ceae67-074a-4d96-9867-f9ce6adc80d3.png)

`CodeEdit` seems to have issues calculating the gutter. Breakpoints and bookmarks lay over each other, and bookmarks are cropped for some reason (the icon is in fact wider, almost square, and has round corners).

![godot windows tools 64_2021-12-31_20-26-37](https://user-images.githubusercontent.com/11782833/147834520-949a23a9-1a3c-48e7-a4d0-b4b2e5b4554e.png)

Something is wrong with `Tree` and editable items. After focusing on the item for a few moments, an overlay appears over it, and the editable state is no longer accessible. Happens with every editable cell that has line edit, at least.

Calinou commented:
I pushed a commit that includes @pycbouh's squashed changes.

akien-mga commented on scene/resources/default_theme/default_theme_icons_builders.py:
```diff
@@ -0,0 +1,77 @@
+    s.write("#ifndef _DEFAULT_THEME_ICONS_H\n")
+    s.write("#define _DEFAULT_THEME_ICONS_H\n")
+    s.write("static const int default_theme_icons_count = {};\n\n".format(len(svg_icons)))
+    s.write("#ifdef MODULE_SVG_ENABLED\n")
```
You need to include `modules/modules_enabled.gen.h` to know which modules are enabled, we don't pass the values to the main env to avoid bloating the command line with too many defines.
```suggestion
    s.write('#include "modules/modules_enabled.gen.h"\n')
    s.write("#ifdef MODULE_SVG_ENABLED\n")
```
(or whatever string formatting `black` would use to allow inner quotes)

akien-mga commented on scene/resources/default_theme/default_theme.cpp:
```diff
@@ -32,97 +32,62 @@
-#include "theme_data.h"
 
-typedef Map<const void *, Ref<ImageTexture>> TexCacheMap;
+#ifdef MODULE_SVG_ENABLED
+#include "modules/svg/image_loader_svg.h"
+#endif
```
```suggestion
#include "modules/modules_enabled.gen.h" // For svg.
#ifdef MODULE_SVG_ENABLED
#include "modules/svg/image_loader_svg.h"
#endif
```

akien-mga commented:
You'll also need to sync the class reference next time you rebase.

pycbouh commented:
The lights are green! I think it's good to go

akien-mga commented:
Thanks!
